src/main/scala/org/allenai/pipeline/Pipeline.scala

```diff
@@ -66,27 +96,41 @@ trait Pipeline extends Logging {
     io: ArtifactIo[T, A],
     fileName: Option[String] = None,
     suffix: String = ""
-  ) = {
-    val path = fileName.getOrElse {
-      // Although the persistence method does not affect the signature
-      // (the same object will be returned in all cases), it is used
-      // to determine the output path, to avoid parsing incompatible data
-      val signature = original.stepInfo.copy(
-        dependencies = original.stepInfo.dependencies + ("io" -> io)
-      ).signature
-      s"${signature.name}.${signature.id}$suffix"
-    }
-    implicitly[ClassTag[A]].runtimeClass match {
-      case c if c == classOf[FlatArtifact] => original.persisted(
-        io,
-        artifactFactory.flatArtifact(s"data/$path").asInstanceOf[A]
-      )
-      case c if c == classOf[StructuredArtifact] => original.persisted(
-        io,
-        artifactFactory.structuredArtifact(s"data/$path").asInstanceOf[A]
-      )
-      case _ => sys.error(s"Cannot persist using io class of unknown type $io")
-    }
+  ): PersistedProducer[T, A] = {
+    makePersisted(original, io, fileName.getOrElse(signaturePath(original, io, suffix)), artifactFactory)
+  }
+
+  protected def signaturePath[T, A <: Artifact](p: Producer[T], io: ArtifactIo[T, A], suffix: String) = {
+    // Although the persistence method does not affect the signature
+    // (the same object will be returned in all cases), it is used
+    // to determine the output path, to avoid parsing incompatible data
+    val signature = p.stepInfo.copy(
+      dependencies = p.stepInfo.dependencies + ("io" -> io)
+    ).signature
+    s"${signature.name}.${signature.id}$suffix"
+  }
+
+  protected def makePersisted[T, A <: Artifact: ClassTag](
```

Comment:
Method blurb please? It makes it easier to understand the intent of this method.